proposals/text/0000-shaper-in-prompt-template.md

```diff
@@ -0,0 +1,153 @@
+ 'node_id': 'PromptNode'}
+    ```
+
+The corresponding prompt template would look like this (provided `value_to_list` and `join_documents` Shaper functions are extended a bit):
+
+    ```python
+    PromptTemplate(
+            name="question-answering-with-references",
+            prompt_text="Create a concise and informative answer (no more than 50 words) for a given question "
+            "based solely on the given documents. You must only use information from the given documents. "
+            "Use an unbiased and journalistic tone. Do not repeat text. Cite the documents using Document[number] notation. "
+            "If multiple documents contain the answer, cite those documents like ‘as stated in Document[number,number,etc]’. "
+            "If the documents do not contain the answer to the question, say that ‘answering is not possible given the available information.’\n"
+            "$documents \n Question: $query; Answer: ",
+            input_shapers=[
+                Shaper(
+                    func="join_documents",
+                    inputs={
+                        "documents": "documents",
+                        "delimiter": "delimiter",
+                        "pattern": "pattern",
+                        "str_replace": "str_replace",
+                    },
+                    outputs=["documents"],
+                    params={
+                        "delimiter": "\n",
+                        "pattern": "\nDocument[$idx]: $content",
+                        "str_replace": {"\n": " ", "[": "(", "]": ")"},
+                    },
+                ),
+                Shaper(func="value_to_list", inputs={"value": "query", "target_list": "documents"}, outputs=["query"]),
+            ],
+            output_shapers=[
+                Shaper(
+                    func="strings_to_answers",
+                    inputs={"strings": "results", "documents": "documents"},
+                    outputs=["answers"],
+                )
+            ],
+            output_variable="answers",
+        )
+    ```
+
+As it's crucial to iterate quickly on prompts, shaper params can be overridden by PromptNode like this:
```

Comment:
Could we also add `shaper_params` to the `run` or `prompt` method as well? That way we could change this at runtime without needing to reload the whole `PromptNode`, which could be slow if we were to use the flan-T5 models.

Reply:
Very good point. 🚀

Reply:
Sure, however for model reuse PromptModel is the way to go, no?

Reply:
Yes @tstadel that is a good point. I think having both ways for being able to experiment with PromptNode without needing to reload the model would be nice.